Feature/update call command

codespool opened the pull request:
Replaced `cargo contract` process spawn with a call to polkadot.js

- added new required argument to differentiate query and tx messages
- dry run skips the tx calling
- renamed args to params or `-p`, and it can take 0 or more parameters (space delimited)

shunsukew commented on packages/cli/src/commands/contract/call.ts:
```diff
@@ -57,17 +95,94 @@ export class CallContract extends Command {
+        );
+
+    const networkUrl = resolveNetworkUrl(config, flags.network ?? "");
+    console.log("AAAA", flags.params);
```
Perhaps you forgot to delete this?

shunsukew commented on packages/cli/src/commands/contract/call.ts:
```diff
@@ -57,17 +95,94 @@ export class CallContract extends Command {
+        );
+
+    const networkUrl = resolveNetworkUrl(config, flags.network ?? "");
+    console.log("AAAA", flags.params);
```
Perhaps you forgot to delete this?

codespool replied:
of course 😓 thanks

shunsukew commented on packages/cli/src/commands/contract/call.ts:
```diff
@@ -1,49 +1,87 @@
+      char: "p",
     }),
-    contractName: Flags.string({
+    account: Flags.string({
```
I think for query call, we can use Alice by default without requiring users to input.

shunsukew commented on packages/cli/src/commands/contract/call.ts:
```diff
@@ -57,17 +95,94 @@ export class CallContract extends Command {
+
+    if (args.messageType === "query") {
+      console.log(`Query result:`);
+      console.log(queryResult.result.toHuman());
```
For now, it is okay, but this needs to be more human understandable, extracting necessary information only for example.

codespool replied:
agreed, we can discuss how to parse it for the next version.

shunsukew commented on packages/cli/src/commands/contract/call.ts:
```diff
@@ -1,49 +1,87 @@
       char: "n",
       description: "Network name to connect to",
     }),
-    deploymentTimestamp: Flags.integer({
```
We lost the way to choose which contracts to call in case several deployments we have.
So, we need to recover some alternatives.

I suggest `contractAddress` or `contractAccountId`, because it's universal concept in blockchain space.
And users don't need to convert something (e.g. timestamp) to a contract address in their head when trying to look up in explorers or some external tools to check.

codespool replied:
Added `address` flag in the update

shunsukew commented:
In case flipper `get` query result is `false`. data is `0x00` it's correct but not friendly.
<img width="510" alt="Screen Shot 2023-01-03 at 12 24 15" src="https://user-images.githubusercontent.com/28953860/210297488-829c736e-a01d-4e0f-8cc7-4e112fc3d696.png">

flipper `flip` tx result. it succeeds but showing error field null is a bit confusing at first sight.
<img width="663" alt="Screen Shot 2023-01-03 at 12 23 59" src="https://user-images.githubusercontent.com/28953860/210297573-886f71c8-6d9e-409f-b677-213a4517fca1.png">
flipper `flip` tx result

codespool commented:
While resolving and updating the flags, I decided to fix it properly, so I split the command into `query` and `tx`, and moved all the shared functionality to a parent class `contractCall`.
Later, the `deploy` command can also use it.

Also, I've added a top level base class, which checks and loads swanky.config, so the rest of the commands need to be updated to use it in the future as well.

shunsukew commented on packages/core/src/lib/nodeInfo.ts:
```diff
@@ -1,13 +1,13 @@
-  supportedInk: "v3.3.1",
+  version: "1.0.0",
+  polkadotPalletVersions: "polkadot-v0.9.33",
+  supportedInk: "v4.0.0",
```
I think if we update supported ink version, better to update templates at the same time as well. Although I assume we can deploy ink 3.x.x contracts to swanky-node 1.0.0 (technically not all, some doesn't have backward compatibility).
Plus, we have UI issue, polkadot.js is available for WeightV1 node but contracts-ui is not available for WeightV2 swanky-node 1.0.0 due to instant finalization issue. So, I suggest to update local node with a separated PR at the right timing.

shunsukew replied:
Also, for weightV2 node, we need to construct gasLimit like this https://github.com/polkadot-js/api/issues/5255#issuecomment-1372540513

shunsukew reviewed (changes requested):
When flipper contract deployed to Shibuya,
Query command gets error for some reason, need investigation.

Shibuya
<img width="1159" alt="Screen Shot 2023-01-20 at 12 23 56" src="https://user-images.githubusercontent.com/28953860/213611264-52ad16bd-c65a-41f3-b5e3-98ae90f33715.png">

swanky-node v0.13.1
<img width="1148" alt="Screen Shot 2023-01-20 at 12 24 37" src="https://user-images.githubusercontent.com/28953860/213611301-80344913-a017-4dba-a872-899cc88c5b9e.png">

shunsukew commented on packages/cli/src/commands/contract/tx.ts:
```diff
@@ -0,0 +1,73 @@
+      ...flags.params
+    );
+    await txResult.signAndSend(this.account.pair, (result) => {
+      if (result.status.isInBlock) {
```
```suggestion
      if (result.status.isInBlock || result.status.isFinalized) {
```

We have instant finalization, when getting isFinalized instead of isInBlock, cli shows nothing and pause.
<img width="1148" alt="Screen Shot 2023-01-20 at 12 24 37" src="https://user-images.githubusercontent.com/28953860/213612960-78200305-e814-498b-8ba9-95c01fe62e23.png">

shunsukew commented on packages/cli/src/commands/contract/query.ts:
```diff
@@ -0,0 +1,32 @@
+    );
+
+    console.log(`Query result:`);
+    console.log(queryResult.result.toHuman());
```
Process doesn't exit properly after query call.
Connection persists?
<img width="1106" alt="Screen Shot 2023-01-20 at 14 33 21" src="https://user-images.githubusercontent.com/28953860/213624703-2b09962d-9392-4219-92f5-288cb4fa261e.png">

shunsukew commented on packages/cli/src/commands/contract/tx.ts:
```diff
@@ -0,0 +1,76 @@
+      ...flags.params
+    );
+
+    this.log(`Gas required: ${queryResult.gasRequired.toHuman()}`);
```
When getting weightV2, log shows `[object Object]`.
If this is weight V2, better to show refTime and proofSize.
<img width="1146" alt="Screen Shot 2023-01-23 at 10 31 04" src="https://user-images.githubusercontent.com/28953860/213952395-342ad9f8-1034-4514-8da7-29777db28ddf.png">
